### R01.6_MergeChroms_RunSLIM_SubsetDMS.R ###

########################################################################################################################################

# Created by: Charley
# Date: 21st Mar 2024
# For Evol Apps

# Merge PQLSeq outputs by chromosome back together into whole genome file
# Apply SLIM multiple testing correction for p-values generated by PQLseq
# Remake methylKit methDiff object from output and save DMS

########################################################################################################################################

###############################
###### Prep environment #######
###############################

###### Load packages ######

library(tidyverse)
library(methylKit)
library(limma)

####### Set directories ######

CHROM_DIR <- "/data/SBCS-EizaguirreLab/Turtle_WGBS/05_MethylKit/04_Evol_Apps_Project/01_Working/02_MethylKit_Objects/02_PQLseq_Objects/Output_PQLseq/By_Chrom/Transform"
PQL_DIR <- "/data/SBCS-EizaguirreLab/Turtle_WGBS/05_MethylKit/04_Evol_Apps_Project/01_Working/02_MethylKit_Objects/02_PQLseq_Objects/Output_PQLseq"
MDIFF_DIR <- "/data/SBCS-EizaguirreLab/Turtle_WGBS/05_MethylKit/04_Evol_Apps_Project/01_Working/02_MethylKit_Objects/02_MethDiff_Objects"
DMS_DIR <- "/data/SBCS-EizaguirreLab/Turtle_WGBS/05_MethylKit/04_Evol_Apps_Project/01_Working/02_MethylKit_Objects/03_DMS_Objects"
UniteCov_DIR <- "/data/SBCS-EizaguirreLab/Turtle_WGBS/05_MethylKit/04_Evol_Apps_Project/01_Working/02_MethylKit_Objects/01_UniteCov_Objects"

########################################################################################################################################

#######################################
###### Merge all chrom together #######
#######################################

List_fit <- list()
List_dfs_fit <- list() 

# Run loop adding file for each chromosome into list of dfs
# NB. R cannot handle starting loop on 0 -> need to bypass
# Create list of chrom, with 0 at end to match order in ref genome - solves object issues later: has to be in same
# order as ref genome
chrom_number <- 1:28
chrom_number <- append(chrom_number, '0', after=28)
chroms <- c(paste0("chr", chrom_number))

prefix <- "PQLseq_fit_" # File prefix

for (chr in chroms) {
  print(chr)
  List_fit[[ chr ]] <- readRDS(file.path(CHROM_DIR, paste0(prefix, chr, ".RDS")))
  List_dfs_fit[[ chr ]] <- List_fit[[ chr ]]
}

# Bind together into a data frame
df_AllChrms <- do.call(rbind, List_dfs_fit)

# % of NaNs for h2
length(df_AllChrms$h2[grep("NaN", df_AllChrms$h2)])/nrow(df_AllChrms)*100 # 17.3% are NaNs, 49.8% for transformed

# For some reason, chr<n> has been added to start of every row name again -> remove
# rownames(df_AllChrms) <- gsub("^.*?\\.","", rownames(df_AllChrms)) # Remove everything before first period

# Save object
saveRDS(df_AllChrms, file.path(PQL_DIR, "PQLseq_Fit_Whole_Genome.RDS"))

########################################################################################################################################

##################################
###### Run SLIM correction #######
##################################

# SLIMcodes.R downloaded from: https://www.aspendb.org/downloads 

###### Define SLIM functions ###### 

# Nothing changed from default in SLIMcodes.R (methylKit also uses default settings)

SLIMfunc<-function(rawp,STA=.1,Divi=10,Pz=0.05,B=100,Bplot=FALSE)
{
  ####################
  m <- length(rawp) 
  
  ########################
  alpha_mtx=NULL;#
  pi0s_est_COM=NULL;
  SzCluster_mtx=NULL;
  P_pi1_mtx=NULL;
  pi1_act_mtx=NULL;
  Dist_group=NULL;
  Num_mtx=NULL;
  Gamma=NULL;
  PI0=NULL;
  
  #############
  ##observed points
  lambda_ga=seq(0,1,0.001);
  gamma_ga=sapply(lambda_ga,f1,rawp=rawp);
  Gamma=c(Gamma,gamma_ga);
  alpha_mtx=c(alpha_mtx,gamma_ga[which(lambda_ga==0.05)]);
  
  
  ###esimation
  pi0_mtx=NULL;
  x.axis=NULL;
  y.axis=NULL;
  itv=(1-STA)/Divi;
  for (i in 1:Divi)##10 for uniform data
  {
    cutoff=STA+(i/Divi)*(1-STA);##10 for uniform data
    lambda=seq(cutoff-itv,cutoff,itv/10);
    gamma_mtx=sapply(lambda,f1,rawp=rawp);
    LModel=lm(gamma_mtx~lambda);
    pi0_mtx=c(pi0_mtx,coefficients(LModel)[2]);
  }
  
  
  ##################################
  ########searching
  N_COM=NULL;
  N_rawp=NULL;
  maxFDR_mtx=NULL;
  quapoint_mtx=NULL; 
  if (B<=1) B=100;
  quapoint_mtx=seq(0.01,0.99,1/B);
  for (k in 1:length(quapoint_mtx))
  {
    qua_point=quapoint_mtx[k];
    
    pi0_combLR=min(quantile(pi0_mtx,qua_point),1);#mean(pi0_mtx);#median();# qua_point=0.78 for desreasing distribution;
    ##0.4 for uniform or normal distribution;
    pi0_est=pi0_combLR;
    
    ###########Calculate independent index of raw p vlaues
    PI0=rbind(PI0,pi0_mtx);
    
    pi0s_est_COM=c(pi0s_est_COM,pi0_est);
    ##Condition1
    P_pi1=sort(rawp)[max(length(rawp)*(1-pi0_est),1)];##
    P_pi1_mtx=c(P_pi1_mtx,P_pi1);
    
    pi0=pi0_est;
    if (is.null(Pz)) Pz=0.05;
    maxFDR=Pz*pi0/(1-(1-Pz)*pi0);
    maxFDR_mtx=c(maxFDR_mtx,maxFDR);
    
    qvalues_combLR=QValuesfun(rawp,pi0);
    qvalue_cf=maxFDR;
    selected=which(qvalues_combLR<qvalue_cf);
    Sel_qvalues_combLR=selected;
    
    pi1_act_mtx=c(pi1_act_mtx,length(Sel_qvalues_combLR)/length(rawp));
    N_COM=c(N_COM,list(Sel_qvalues_combLR));
    Num_mtx=c(Num_mtx,length(Sel_qvalues_combLR));
  }
  length(N_COM)
  length(quapoint_mtx)
  
  ####doing judging
  ##by max FDR
  pi1s_est_COM=1-pi0s_est_COM;
  Diff=sum(rawp<=Pz)/length(rawp)-pi1_act_mtx;
  
  ###
  loc=which.min(abs(Diff));
  Diff.loc=Diff[loc];
  selQuantile=quapoint_mtx[loc];
  pi0_Est=min(1,pi0s_est_COM[loc]);
  maxFDR.Pz=Pz*pi0_Est/(1-(1-Pz)*pi0_Est);
  
  if(Bplot)
  {
    windows();
    par(mfrow=c(1,2));
    hist(rawp,main="Histogram of p-value");
    gamma_ga=sapply(lambda_ga,f1,rawp=rawp);
    plot(lambda_ga,gamma_ga,type="l",main="Relationship of p- and q value",xlab=expression(lambda),ylab=expression(gamma),cex.lab=1.45,cex.axis=1.42)
    #par(xaxp=c(0,1,10));
    #axis(1);
    ##qvalues
    qValues=QValuesfun(rawp,pi0=pi0_Est);
    gammaq_ga=sapply(lambda_ga,f1,rawp=qValues);
    lines(lambda_ga,gammaq_ga,col="blue",lwd=2,lty="dashed")
    abline(v=Pz,col="black",lwd=2,lty="dotdash")
    abline(v=maxFDR.Pz,col="blue",lwd=2,lty="dotdash")
    text(0.75,0.6,labels=paste("L=",round(abs(Diff.loc),4),sep=""));
    leg=list(bquote("CPD of p-value"),bquote("CPD of q-value"),bquote("Pmax"==.(Pz)),bquote("FDRmax"==.(round(maxFDR.Pz,2))));
    legend("bottomright",legend=as.expression(leg),lwd=2,lty=c("solid","dashed","dotdash","dotdash"),col=c("black","blue","black","blue"));
  }
  
  return(list(pi0_Est=pi0_Est,selQuantile=selQuantile));
}

f1 <- function(cutoff,rawp){sum(rawp<cutoff)/length(rawp)*100};

###
# Define function to calculate q-values
QValuesfun<-function(rawp,pi0)
{
  order_rawp=sort(rawp);
  qvalues=pi0*length(order_rawp)*order_rawp/c(1:length(order_rawp));
  temp=cummin(qvalues[seq(length(qvalues),1,-1)])
  qvalues=temp[seq(length(temp),1,-1)];
  qvalues=qvalues[order(order(rawp))]
}


###### Run SLIM ######

df_AllChrms <- readRDS(file.path(PQL_DIR, "PQLseq_Fit_Whole_Genome.RDS"))

pvals <- df_AllChrms$pvalue
length(pvals)

# Run SLIM
qvals = {QValuesfun(pvals,
                 SLIMfunc(pvals,STA=.1,Divi=10,Pz=0.05,B=100,Bplot=FALSE)$pi0_Est)
}

# Add to fit table
slim_output <- df_AllChrms
slim_output$qvalue <- qvals

###### Checks ######

# How many pass cut off?
sum(slim_output$pvalue < 0.01) # 32141
sum(slim_output$qvalue < 0.01) # 4330
sum(slim_output$pvalue < 0.01)/nrow(slim_output)*100 # 1.175787%
sum(slim_output$qvalue < 0.01)/nrow(slim_output)*100 # 0.1584007%

# Summary
summary(slim_output$qvalue)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.0000  0.9984  0.9984  0.9902  0.9984  0.9984 

hist(slim_output$pvalue)
hist(slim_output$qvalue)

# Save
saveRDS(slim_output, file.path(PQL_DIR, "PQLseq_Fit_Transform_SLIM_Corrected_Whole_Genome.RDS"))


########################################################################################################################################

#######################################################
###### Combine with methylation difference info #######
#######################################################

# methylKit::CalculateDiffMeth generates calculation of methylation difference. This is simply the mean % meth of group 1 - mean % meth of group 2
# -> not affected by statistical model. 
# We will just replace p and q values, as useful to have object in methylKit MethDiff format

slim_output <- readRDS(file.path(PQL_DIR, "PQLseq_Fit_SLIM_Corrected_Whole_Genome.RDS"))
methdiff <- readRDS("/data/SBCS-EizaguirreLab/Turtle_WGBS/05_MethylKit/04_Evol_Apps_Project/01_Working/02_MethylKit_Objects/02_MethDiff_Objects/EvolApps_WholeGenome_ALL_MethDiff_filt_CtoT_Intersect.RDS")

####### Create a new methylDiff object ######

# NB. Get objects into the same order first to be able to detect matches

### Prep methylDiff object ###
methdiff_df <- as.data.frame(getData(methdiff)) # Turn methylDiff object into a df: need getData() otherwise doesn't coerce
class(methdiff_df)
methdiff_df <- methdiff_df %>% dplyr::select( -pvalue, -qvalue) # Subset for 1st 4 columns
methdiff_df$siteloc <- paste(methdiff_df$chr, methdiff_df$start, sep = ".") # Add genomic location column

### Prep SLIM output ###
slim_output$siteloc <- rownames(slim_output)  # Add genomic location column

### Reorder dfs so in same genomic order ###
# Rows are in different orders in the 2 dfs
# Save indices for how to reorder second to match first
reorder_idx <- match(methdiff_df$siteloc, slim_output$siteloc) # Save indices for how to reorder slim_output to match methdiff_df
sum(is.na(reorder_idx)) # Check
slim_output_reorder <- slim_output[reorder_idx,] # Apply
identical(rownames(methdiff_df), rownames(slim_output_reorder)) # Check: TRUE

### Merge ###
# Add p and q value from PQLseq, before meth.diff so it's in the same position
merged_df <- add_column(methdiff_df, pvalue = slim_output_reorder$pvalue, .before = "meth.diff")
merged_df <- add_column(merged_df, qvalue = slim_output_reorder$qvalue, .before = "meth.diff")

# Remove siteloc
merged_df$siteloc <- NULL

### Convert back into methylDiff object by adding metadata ###
new_methdiff <- new(Class = "methylDiff", merged_df,
                    sample.ids=methdiff@sample.ids,
                    assembly=methdiff@assembly,
                    context=methdiff@context,
                    treatment=methdiff@treatment,
                    destranded=methdiff@destranded,
                    resolution=methdiff@resolution,
                    names=methdiff@names,
                    row.names=rownames(merged_df) # New row names with chr.pos
)

# Save
saveRDS(new_methdiff, file.path(MDIFF_DIR, "EvolApps_WholeGenome_ALL_MethDiff_filt_CtoT_Intersect_PQLseq.RDS"))

########################################################################################################################################

#########################
###### Subset DMS #######
#########################

new_methdiff <- readRDS(file.path(MDIFF_DIR, "EvolApps_WholeGenome_ALL_MethDiff_filt_CtoT_Intersect_PQLseq.RDS")) # PQLSeq
methdiff <- readRDS(file.path(MDIFF_DIR, "EvolApps_WholeGenome_ALL_MethDiff_CovarMat_filt_CtoT_Intersect.RDS")) # methylKit, to compare against

PQL_DMS_10pc_q0.05 = getMethylDiff(new_methdiff, difference=10, qvalue=0.05)

# Filter out all q=0 sites from uniteCov object
# These are excluded after visual exploration, as they represent non-meaningful values where PQLseq algorithm has likely failed
# at methylation boundaries of 0 and 1. NB. Checked results with DSS, as it uses arcsine function to better deal with this. Confirmed
# that q=0 values from PQLseq are non meaningful, and would have been filtered out anyway in DSS with high p value
PQL_DMS_10pc_q0.05_no0q <- as.data.frame(getData(PQL_DMS_10pc_q0.05))
PQL_DMS_10pc_q0.05_no0q <- PQL_DMS_10pc_q0.05_no0q[PQL_DMS_10pc_q0.05_no0q$qvalue > 0,] # 287 DMS
saveRDS(PQL_DMS_10pc_q0.05_no0q, file.path(DMS_DIR, "EvolApps_WholeGenome_ALL_DMS_PQLseq_10pc_q0.05_no0q.RDS"))

###### Create uniteCovDM file for DMS ######

PQL_DMS_10pc_q0.05_no0q <- readRDS(file.path(DMS_DIR, "EvolApps_WholeGenome_ALL_DMS_PQLseq_10pc_q0.05_no0q.RDS"))
uniteCovALL <- readRDS(file.path(UniteCov_DIR, "EvolApps_WholeGenome_uniteCovALL_filt_CtoT_Intersect.RDS"))

# Create vector of DMS with just contig name and position in contig
myPos = paste(PQL_DMS_0pc_q0.05$chr, PQL_DMS_0pc_q0.05$end)
# Change the Pos vector into a dataframe): called df
df <- data.frame(chr=sapply(strsplit(myPos, " "), `[`, 1),
                 start=sapply(strsplit(myPos, " "), `[`, 2),
                 end=sapply(strsplit(myPos, " "), `[`, 2))

# Create GRanges object from DMS df
GRanges_PQL_DMS <- makeGRangesFromDataFrame(df)

# Selects records that match list of DMS locations
uniteCovDM <- selectByOverlap(uniteCovALL,GRanges_PQL_DMS)

# Save 
saveRDS(uniteCovDM, file.path(UniteCov_DIR, "EvolApps_WholeGenome_uniteCovDM_ALL_PQLSeq_0pc_q0.05.RDS"))
